Implementing additional Query-level retry logic

victor-torres opened the pull request:
This retry logic was originally proposed by @ivanprado. 

I've refactored the source code moving most part of the logic to the `QueryError` exception. 
I've also introduced the sub-batches retry.

ivanprado commented on .travis.yml:
```diff
@@ -8,8 +8,6 @@ branches:
 
 matrix:
   include:
-    - python: 3.5
```
Why are we removing support for python 3.5? Is such is the case we have to update the README as well.

victor-torres replied:
Type hints are heavily used in this project, but they're not supported by Python 3.5 (the way they're used here). Instead of refactoring, it seems like more reasonable to ditch Python 3.5. It also doesn't offer support for f-strings. I'm opening another pull request to make this change alone.

victor-torres replied:
I have removed the Python 3.5 support because of usages of Type Hints in the `autoextract.aio` module were not supported by this version.

I can see in the README file that we're offering support for basic / synchronous calls with Python 3.5. This is weird to because part of the codebase supports a given version of Python while other doesn't. Shouldn't we remove the Python 3.5 support?

Python 3.5 could become difficult to maintain regarding the limitations with Type Hints and they also don't support f-strings.

I guess it should be done in another pull request, tho. I'll try to re-enable Python 3.5 pipelines and skip the incompatible test cases.

ivanprado commented on autoextract/aio/client.py:
```diff
@@ -135,7 +135,14 @@ async def request():
+    for r in result:
+        if "error" in r:
+            # https://doc.scrapinghub.com/autoextract.html#query-level
+            raise QueryError(r["query"], r["error"])
```
Mmm, now I see the complexity. AutoExtract API supports batching (submit a list of queries in the same request). And you can receive both right results and errors in the response. But we wouldn't want to retry for all of them in the case a single one or a few fails, it would be pretty inefficient and will provide bad results in practice. 

The logic to make it work would require to create a new query only with the failed requests and retry for them. And the policy for retrials will be also much more complex, as you can have different kind of errors, and each with a different policy. 

That makes me think that it makes no sense to support `QueryErrors` in this library. Even if we do it, it won't be efficient. Imagine the user submit 100 queries in a single request, 10 returns with errors, so you start retrying them... and one of them is always a timeout, so you consume all the retrials. It can take a lot of time until you come back to the client. And the client has this request slot occupied. 

It would be much more efficient in this case if the client obtains the first response with the errors, and can pack the 10 retrials with other queries he has enqueued. 

Conclusion: query level retrials should be managed at the client. We might create some tooling in this library to simplify the work in the client.

victor-torres replied:
What about the following?
- retry single queries
- don't retry multiple queries (batches) / log a warning

victor-torres replied:
Quick update: Ivan and I have talked about this situation. Another possibility is that we could create a new standalone method for single queries that have this additional retry logic enabled.

ivanprado replied:
This thread is now outdated. Resolving.

victor-torres commented:
- [ ] Revert Python 3.5 removal
- [ ] Skip test cases when incompatible with Python 3.5
- [ ] Experiment with standalone method for single queries with additional retry support

victor-torres commented:
I talked to @ejulio today. Another option proposed by him is turning the request method into an async generator instead of returning a list of responses. The code would retry only failed responses. We would promptly yield successful ones.

Another modification that could take place is receiving a flag enabling or disabling the additional retry logic for Query-level errors.

victor-torres commented on autoextract/aio/client.py:
```diff
@@ -114,7 +121,31 @@ async def request():
+
+                pending_queries.clear()
+                for query_result in response:
+                    if "error" in query_result:
```
`... and handle_retries `

victor-torres replied:
done

ivanprado commented on autoextract/aio/retry.py:
```diff
@@ -95,13 +105,22 @@ def __call__(self, retry_state: RetryCallState) -> float:
+                logger.warning(exc)
+                return 300  # 5 minutes
+
+            return wait or self.retriable_wait(retry_state=retry_state)
```
I think we should retry on max(wait, self.retriable_wait).

victor-torres replied:
Done.

victor-torres commented on autoextract/aio/retry.py:
```diff
@@ -95,13 +105,22 @@ def __call__(self, retry_state: RetryCallState) -> float:
 class autoextract_stop_strategy(stop_base):
     def __init__(self):
         self.stop_on_throttling_error = stop_never
+        self.stop_on_retriable_query_error = stop_never
```
This is something I wanted to discuss with you guys. Even with the domain occupied errors, I guess it's somewhat dangerous to ask for perpetual retries. Even the 15 minutes default stop time sounds like too much to me. Should we be waiting that long for responses? Another option is setting a max number of retries (e.g. 3 retries).

ivanprado replied:
> Should we be waiting that long for responses?

I think it depends. Let's imagine that you receive a 500. The site could be down, and it could take a while until it comes back to life. 

In any case, probably this should be configurable somehow. Some user might need faster responses so setting a lower timeout should be possible. But I guess this is out of the scope of this ticket. 

> Another option is setting a max number of retries

I prefer using timing and jitter. It would be beneficial for the API.

victor-torres replied:
With last commit I make sure we're following your original implementation:

- retry domain occupied errors forever
- retry other query-level errors up to 15 minutes

victor-torres commented on autoextract/aio/retry.py:
```diff
@@ -95,13 +105,22 @@ def __call__(self, retry_state: RetryCallState) -> float:
+            try:
+                wait = exc.retry_seconds
+            except ValueError as exc:
+                logger.warning(exc)
```
```python
    raise ValueError(
                f"Could not extract retry seconds "
                f"from Domain Occupied error message: {self.message}"
            )
```

victor-torres replied:
Instead of raising an exception, we're logging a warning message and returning a default value (5 minutes).
The code that previously relied on `QueryError.retry_seconds` will no longer need to capture `ValueError` exceptions.

victor-torres commented on autoextract/aio/client.py:
```diff
@@ -114,9 +126,33 @@ async def request():
+                            query_exceptions.append(query_exception)
+                            user_query = query_result["query"]["userQuery"]
+                            pending_queries.append(user_query)
+                            continue
```
Code was updated to return available results (including both successes and errors) when a RetryError exception is raised.

ivanprado commented on autoextract/aio/client.py:
```diff
@@ -61,11 +160,16 @@ async def request_raw(query: Query,
+       if there is a Request-level error returned by the API
+       which is not a throttling response
+       (e.g. it can be raised for incorrect request).
+    2) autoextract.errors.QueryError,
```
I don't think this method can raise QueryError. It shouldn't.

victor-torres replied:
That's true. QueryError exceptions should be catch and handled by the RequestProcessor. I've updated this docstring.

ivanprado commented on autoextract/aio/client.py:
```diff
@@ -77,21 +181,28 @@ async def request_raw(query: Query,
         stats = ResponseStats.create(start_global)
         agg_stats.n_attempts += 1
+
+        kwargs = dict(
```
Nipnick: `url`, `auth` and `headers` can be cached outside the method to avoid recreating them each time.

victor-torres replied:
- `endpoint` (`url`) is already an argument so it's not expected to change during runtime. 
- `auth` and `headers` were moved outside the `request` method; thanks!

ivanprado commented on autoextract/aio/errors.py:
```diff
@@ -15,3 +22,64 @@ def __str__(self):
+        if self.domain_occupied:
+            return True
+
+        return self.message.lower() in self.RETRIABLE_QUERY_ERROR_MESSAGES
```
```suggestion
        return self.message.lower().strip() in self.RETRIABLE_QUERY_ERROR_MESSAGES
```

ivanprado commented on autoextract/aio/errors.py:
```diff
@@ -15,3 +22,64 @@ def __str__(self):
+        return match.group(1) if match else None
+
+    @property
+    def retry_seconds(self) -> Optional[float]:
```
Suggestion. What do you think about creating a class `DomainOccupied`:
* with two members:
  * domain
  * retry_seconds
* That also contains the REGEX and the logic for the match. 
* With a class method to build it, e.g. `from_query_error` that returns None if it is not a domain occupied method. 

And finally, have a single property in `QuerError` `domain_occupied` with return type `Optional[DomainOccupied]`

ejulio replied:
I agree with @ivanprado , but I would only consider it if the `DomainOccupied` logic is used in other places, otherwise, I guess it is fine to keep it here.

victor-torres replied:
I've tried to move the Domain Occupied logic to a dedicated place. This made the QueryError exception class a little bit more concise. I've kept the `retry_seconds` property as a convenience in order to keep retry and client logic more simple (e.g. `_get_exception_with_longest_timeout`). Let me know what you think.

ivanprado replied:
I see that this change implies more lines added that those removed, so not sure anymore. Anyway, I think that also `retry_seconds` should be removed from RequestError, because having it there suggest that every `RequestError` can have some retry seconds, which is not the case. 

But anyway, this is very minor, just a matter of taste as well. So I'm happy with original version or with this one if you want to keep them.

victor-torres replied:
> I see that this change implies more lines added that those removed

Indeed but at least things are a little bit more separated now.

> Anyway, I think that also retry_seconds should be removed from RequestError, because having it there suggest that every RequestError can have some retry seconds, which is not the case.

Not really, it has the same signature as before my refactor. Look that our return type is an **optional float**.

I like `QueryError` having this `retry_seconds` property because it could make it easier to add custom retry seconds depending on the Query-level error type, instead of having to depend on the `retry.py` file.

At the end of the day, I don't have a preference between these two versions neither. Let's wait for @ejulio's input here otherwise, we could resolve this conversation the way it is now.

ejulio replied:
No strong opinion on this one.
If there is no time to retry because it is not retriable, the the Optional makes sense.
If it is because we can run right away, I would simplify and return as 0 instead of `None`, as 0 means no wait time required, while `None` gives the idea of `this is not retriable`

victor-torres replied:
Done.

victor-torres commented on autoextract/aio/client.py:
```diff
@@ -34,6 +35,104 @@ class Result(List[Dict]):
+        """Clear temporary variables between retries"""
+        self.pending_queries = list()
+        self._query_errors = list()
+        self._query_exceptions = list()
```
Done

victor-torres commented on autoextract/aio/client.py:
```diff
@@ -34,6 +35,104 @@ class Result(List[Dict]):
+    def _reset(self):
+        """Clear temporary variables between retries"""
+        self.pending_queries = list()
+        self._query_errors = list()
```
Done

ejulio commented on autoextract/aio/client.py:
```diff
@@ -34,6 +35,104 @@ class Result(List[Dict]):
+        self._reset()
+        for query_result in query_results:
+            if self._handle_retries and "error" in query_result:
+                if self._process_error(query_result):
```
This one is a bit magic, _I'm asking for something (bool), but I'm also changing the class state_.

victor-torres replied:
I've tried to make it less "magical" 😛 

Could you double-check it?

victor-torres replied:
https://github.com/scrapinghub/scrapinghub-autoextract/pull/7/commits/4d632acb2eca4b8845d64ecf434cd0acba6f88aa

ejulio commented on autoextract/aio/client.py:
```diff
@@ -34,6 +35,104 @@ class Result(List[Dict]):
+            self._query_successes.append(query_result)
+
+        if self._query_exceptions:
+            raise self._get_exception_with_longest_timeout()
```
If not used in other places, it could be a variable here.
Probably, would make it a bit clearer

victor-torres replied:
> Probably, would make it a bit clearer

That was my intention when I moved this code to a private class-method but it might have got a little bit more complex and the user needs to refer to this method to understand what's happening here.

Done.

ejulio commented on autoextract/aio/client.py:
```diff
@@ -127,6 +238,15 @@ async def request():
     try:
         result = await request()
+    except RetryError as exc:
+        if isinstance(exc.args[0].exception(), QueryError):
```
It seems this `try` knows a lot on how `request` behaves.
Maybe a specific exception `RetryQueryError` which inherits from `RetryError`, but is more specific and gives more context on what happened?

victor-torres replied:
I've refactored the retry logic to raise a custom/more specific exception when the retry stops because of a `QueryError`. Could you take a look and see if you like it?

ejulio commented on autoextract/aio/errors.py:
```diff
@@ -15,3 +22,64 @@ def __str__(self):
+        return self.message.lower() in self.RETRIABLE_QUERY_ERROR_MESSAGES
+
+    @property
+    def domain_occupied(self) -> Optional[str]:
```
It seems it's used only as a `bool` result, so you can update the typing to be `bool` and return the proper value

victor-torres replied:
Not anymore :)

ejulio commented on tests/test_aio/test_errors.py:
```diff
@@ -0,0 +1,53 @@
+    assert exc.domain_occupied == domain_occupied
+
+    if isinstance(retry_seconds, type) and issubclass(retry_seconds, Exception):
+        with pytest.raises(retry_seconds):
```
Maybe two different functions, so the assertions are clear and do not depend on the argument type?

victor-torres replied:
Will get back to tests when the actual code looks good for all reviewers. Thanks.

ivanprado commented on autoextract/aio/errors.py:
```diff
@@ -8,6 +8,38 @@
+        self.retry_seconds = retry_seconds
+
+    @classmethod
+    def from_message(cls, message: str):
```
```suggestion
    def from_message(cls, message: str) -> Optional["DomainOccupied"]:
```

victor-torres replied:
Done

ejulio commented on autoextract/aio/client.py:
```diff
@@ -34,6 +34,90 @@ class Result(List[Dict]):
+            if self._handle_retries and "error" in query_result:
+                query_exception = QueryError.from_query_result(query_result)
+                if query_exception.retriable:
+                    self._process_error(query_result, query_exception)
```
Just not sure about the name, but I wouldn't bother too much about it.
Maybe `enqueue_errors`?

victor-torres replied:
Renamed

ejulio commented on autoextract/aio/errors.py:
```diff
@@ -1,11 +1,52 @@
+        try:
+            retry_seconds = float(match.group(2))
+        except ValueError:
+            logger.warning(
```
I always have issues when deciding the proper log level.
Lately, I've been trying to use `warning` as something that we could act on.
Of course, if the regex is wrong, we'll get an warning, but on the main case, should it be a warning or simply info/debug?

victor-torres replied:
Perhaps it's closer to a warning because it would require:
- a fix on the API;
- or a fix in our retry seconds parser

kmike commented on autoextract/aio/retry.py:
```diff
@@ -42,7 +43,7 @@
 def _is_network_error(exc: Exception) -> bool:
-    if isinstance(exc, ApiError):
+    if isinstance(exc, RequestError):
         # ApiError is ClientResponseError, which is in the
```
this comment needs to be updated

victor-torres replied:
Done

ivanprado commented on autoextract/aio/errors.py:
```diff
@@ -1,17 +1,114 @@
+            "Proxy error: ssl_tunnel_error",
+            "Proxy error: msgtimeout",
+            "Proxy error: econnrefused",
+            # Retry errors for AutoExtract API dev server
```
Regarding these errors which are related to internal API dev server... we might just keep them outside of the client library and just deal with them in other parts. Ideally, it would be nice to configure which error you want to retry on, but I think this is out of the scope of this PR. So I think it is Ok to remove them from here.

victor-torres replied:
Removed

victor-torres commented on autoextract/__main__.py:
```diff
@@ -116,6 +118,12 @@ def read_input(input_fp, intype, page_type):
                    choices=["DEBUG", "INFO", "WARNING", "ERROR"],
                    help="log level")
     p.add_argument("--shuffle", help="Shuffle input URLs", action="store_true")
```
@kmike @ivanprado 

I've added these two flags to our command-line utility:

- `--disable-retries`
- `--skip-query-errors`

Since they're flags, they're both false by default, which means:

- `handle_retries=True`
- `skip_query_errors=False`

The last argument has no effect if retries are already disabled. This is documented in the help text and in the request function's docstring.

Let me know your thoughts.

victor-torres replied:
I have noticed that we are enabling retries by default. Should we switch this behavior and replace the first flag by something like `--enable-retries` or simply `--retry`? Should we make this change also in the source code or should we keep the current behavior there?

kmike commented on autoextract/aio/client.py:
```diff
@@ -34,11 +34,101 @@ class Result(List[Dict]):
                       endpoint: str = API_ENDPOINT,
                       *,
                       handle_retries: bool = True,
+                      skip_query_errors: bool = False,
```
I'd expect that if skip_query_errors=True is passed, the returned result doesn't contain query errors, not that these errors are no longer retried. 

Unlike errors which are being retried at the moment, query errors may not have a single best strategy for retrial: how much to retry proxy errors, etc. depends on 

* behavior of remote websites
* willingness of user to keep retrying those, as there are diminishing returns

What do you think about having a different argument, something like `query_error_max_retries`? 0 disables retries of query-level errors, other values allow to express willingness of the user to get all results, as compared to costs.
Alternatively, instead of an integer, it can be a tenacity stop condition object, though I don't have a strong opinion on that; int is probably fine.

kmike replied:
If we go with `query_error_max_retries` (or something like that), CLI may need just this option. I'm not sure there is a need to provide an option to disable retries completely. E.g. 429 messages are a part of the protocol, not errors.

victor-torres replied:
I like the max retry count setting, especially if it's an integer because it's easy to implement, easy to understand, and consistent across multiple use cases. It seems like this is the type of argument used by similar API wrapper libraries that need to implement a retry strategy.

I don't think those libraries make any differentiation between paid and free requests, but we should answer this question thinking about our use case: should we differentiate paid from free requests/retries? 

Also, should different Query-level errors have a different max retry count? Should we expand this logic to the Request-level errors as well? Using the same argument or exclusive ones?

If you think this is the path we should follow, I could work on it :) 
Let's discuss about it.

ivanprado replied:
Let me provide my thoughts on all that with a little bit of perspective. This is what theoretically I think the library should support:
* Provide meaningfully query-level errors. This means several things:
  - Have a catalogue of all the possible errors that can be received with a clear explanation about what they mean, when they happen, and which is the recommended action on them (e.g. should they be retried, and how should the be retried). Ideally, I would enumerate these errors so that we can reference them. 
  - This catalogue should be complete enough.
  -  Have groups of these errors. Some example groups I'm thinking about: `retriables`, `retriables-with-suggested-time`, `hard-errors`. There can be more, for example, if we need different retry policies by them. 
* Fine-grained support of customization of the retrial policy in the library
  - Being able to change the policy for a particular query-level error
  - Being able to change the policy for a group of errors
  - Being able to configure the behaviour of the retrials. IMO, it would be enough with three parameters for most of the errors: `max-retrials`, `time-between-retrials`, `max-total-time`.
* Command line should have sensible parameters by default, and probably it won't be required so fine-grained configuration support. Probably enough allowing configuring some general parameters.

The former is the ideal state I think the library should get. But I know we are far from that (especially starting with the catalogue of errors, for which we don't have enough knowledge now and is likely to change in the future). And is out of the scope of this PR to reach to that level.

IMO all steps in this PR that get us closer to this desired state are good to me. `query_error_max_retries` is reasonable to me.

kmike replied:
@victor-torres 

> I don't think those libraries make any differentiation between paid and free requests, but we should answer this question thinking about our use case: should we differentiate paid from free requests/retries?

After looking at exceptions, I think that can be fine not to differentiate them. I think we could differentiate errors which are caused by remote websites or proxies, vs errors which are caused by the API server. All retriable request-level and network errors are caused by the API server, and all retriable query-level errors are caused by remote websites or proxies. So I think that's fine to have logic separated as in this PR: have default, non-customizable behavior for request-level and network errors (as now), and add retries for query-level errors, with a max_retry option.

> Also, should different Query-level errors have a different max retry count? Should we expand this logic to the Request-level errors as well? Using the same argument or exclusive ones?

I think that's fine to leave request-level errors as-is. We should know the best values, because we control the behavior of the API server, and request-level errors and network errors are originated from the API server. 

Retriable query-level errors come from website and from proxies, and so best values can vary, based on websites and on user needs, so it'd be good to allow customization. I think a single max retry value is enough for now, and it can be made more complex later, if we find a need for it.

> I'd go with with a retry count for simplicity, but leave it open for something more "specialized", like tenacity conditions to allow some behavior as described by @ivanprado

+1. We need to do nothing to leave it open for something more specialized, as we can always add a new argument, or reuse the existing one, allowing to receive some tenacity objects.

victor-torres replied:
@kmike just to be sure before getting back to code:

- should I implement max retry count and **keep** current tenacity logic as well?
- if the answer to the question above is positive, should I stop to whatever happens first (max number of retries or tenacity stop time), or is there a priority when using those configurations?
- what should be the default number of retries?

kmike replied:
> should I implement max retry count and keep current tenacity logic as well?

Let's review the current logic.

1. For request-level errors and network errors I think max retry count shouldn't have any effect. I.e. retries which were in a library before this PR should stay the same; the shouldn't depend on max retry parameter.
2. For domain occupied errors there is currently wait time identified in a message + stop_never. I think it should be changed to "stop after max retry count". Ideally, wait time should be picked up from the error message every time, but I'm not sure this is a 100% requirement. If that's hard to implement, using the first value can be ok.
3. For query level errors there is an increasing retry time with jitter + max retry time limit. I'm not sure max retry time limit makes sense if we have max retry time; we just need to pick a good strategy for increasing time between retries. I think it shouldn't be purely exponential, there should be a cap on max time between retrials.

> if the answer to the question above is positive, should I stop to whatever happens first (max number of retries or tenacity stop time), or is there a priority when using those configurations?

I think max number of retrials could be implemented as a part of tenacity stop condition. Regarding logic - see the proposal above.

> what should be the default number of retries?

My vote is for 0.

kmike commented on autoextract/aio/client.py:
```diff
@@ -147,6 +259,8 @@ def request_parallel_as_completed(query: Query,
                                   n_conn=1,
                                   agg_stats: AggStats = None,
+                                  handle_retries=True,
+                                  skip_query_errors=False,
```
When we settle on exact API, it'd be good to document new arguments in a docstring

kmike commented on autoextract/aio/retry.py:
```diff
@@ -104,6 +123,8 @@ def __init__(self):
         self.stop_on_network_error = stop_after_delay(15 * 60)
         self.stop_on_server_error = self.stop_on_network_error
+        self.stop_on_retriable_query_error = self.stop_on_network_error
+        self.stop_on_domain_occupied_query_error = stop_never
```
If I'm not mistaken, this can be problematic.

For example, you create a file with 100K URLs to website A, and 100K URLs to many other websites. Then you start a CLI tool, with 40 connections to the API. Over time you start getting "domain occupied" messages for the website A. This all happens inside request_raw function, so if a domain occupied message is received, the connection won't be reused until the retrials are finished. So in this example all processing stops, all connections are busy retrying domain occupied errors for website A, and work for other websites is not being done.

It can be solved e.g. by having a separate component which limits an amount of requests sent in parallel to a domain. But this is not straightforward to implement, and we don't have it in CLI.

There is no need to create a perfect solution here, but my worry is that with the current approach a command which worked before may get stuck now. 

I think that's correct behavior to retry domain occupied errors forever (if the goal is to process all requests). However, there is a difference from e.g. throttling errors, which also use stop_never. If there is a throttling error, there is no other work you can do in the meantime; throttling happens for all connections, and it'll resolve itself in some time. Domain occupied error is specific to a particular query, and you can do other work while waiting for next retry opportunity. Once a domain is occupied, it is easy to get all available connections stuck, having them retry this error just for a single domain.

Unfortunately this applies to other query errors as well. Network/throttling/server errors are not caused by a single website, they are overall system errors, while query errors do depend on the URL being processed. It seems my proposal of doing retrials inside request_raw was not that good in the end :(

If it works in practice, that's probably fine. A similar approach worked for Scrapy retrials for quite a long time; it was not perfect (sometimes spiders get stuck), but workable. Scrapy was not using infinite retries though - with infinite the behavior could have been much worse, with no chance to recover. So, to make domain occupied errors no worse than query errors, I wonder if we should make their retrials bounded as well for now.

What do you think @victor-torres?

victor-torres replied:
I appreciate your thoughts, Mikhail.

I don't like the "stop never" strategy because it's like having a "while true" condition. It could lead to starvation and waste of computational resources. There should be an exit condition. Even with Domain Occupied errors because you never know.

I'm not comfortable with the long thresholds too, having to wait, for example, 15 minutes before giving up. To some extent, I think the "fail fast" strategy could be applied in those cases. For example, we don't need to spend too much time in our spider waiting for failing queries. We could schedule a supplementary job to grab missing items later. The key here is handing responsibility back to the spider. It should be able to make the best decision: to wait for a longer period of time or to give up temporarily and try again later. Of course, we could supply a better API to make it possible, like the max retry count you've suggested in previous comments.

We knew beforehand that trying to implement such a retry strategy with that granularity and also supporting sub-batches would be a challenge. And it will probably never cover 100% of our use cases. Similar libraries have a much simpler retry strategy. I remember one thing that was discussed was that this retry strategy should be used with small batches, probably to minimize issues like the ones you described. Of course, this is something developers should have in mind while creating solutions using this library, and having to remember things or knowing implementation details is not an ideal situation. 

So trying to be a little bit pragmatic here, we have a couple of options to improve this solution without refactoring a significant part of it:

- we could replace the "stop never" strategy by a finite (and safer) alternative
- we could create a separate flag for Domain Occupied errors

There's also the option to use a max retry count, but I'll discuss it in another comment because I think it's a world apart.

Oh, almost forgot that we had internal discussions to remove the batch support from this wrapper since almost no one is making use of it. Will it be useful in the near future? Is it necessary to overcome some API limitations such as rate limits and RPS? Would the previous issues happen if single requests are made per user query?

ivanprado replied:
Domain occupy seems to be the kind of error that should be better handled by the user (e.g. he can decide to retry on it tomorrow). So I'm Ok to set a limit for it and return it back to the user so that It handles it the best way he can. 

Regarding batch, I think is generating more troubles that time it saves. We should reconsider supporting only single requests. It will be more maintainable and with have a better behaviour.

> Similar libraries have a much simpler retry strategy

The problem is that AutoExtract API as it is designed now have a considerable number of failed responses that requires a retrial. Leaving the retrial logic to the user makes the library simpler but at the same time it is less useful to the user given that he needs to spend considerable efforts to implement the retrial logic (or don't do it and get a suboptimal service quality).

ejulio replied:
Leaving here for history.
I see it as two different things.
For a low level library, I expect it to run at most a few retries and return. In general, I don't expect to call a python function and it will loop forever, this is a kind of huge side effect.
On the other hand, for a CLI, I think it is fine to loop forever, as I want the results, no matter the time it takes to get them.
So, maybe a better separation of concerns of what is the AutoExtract integration library (wrapper to HTTP calls) and what is a CLI framework that can be used to extract data from a set of URLs.

kmike replied:
@victor-torres 

>  we could replace the "stop never" strategy by a finite (and safer) alternative
    we could create a separate flag for Domain Occupied errors

+1 to replace stop_never for now, and move ahead. I think we can leave without special logic for Domain Occupied. We can e.g. respect time provided in messages, but reuse the same max-retry amount.

> Oh, almost forgot that we had internal discussions to remove the batch support from this wrapper since almost no one is making use of it. Will it be useful in the near future? Is it necessary to overcome some API limitations such as rate limits and RPS? Would the previous issues happen if single requests are made per user query?

I think the problem described here is not related to batches, and won't be solved by removing batching support.
The issue is cause by the way retries are handled. Currently architecture looks like this (assuming concurrency is set to 3):

```
[List of URLs] 
   -> request_raw #1, retry until success 
   -> request_raw #2, retry until success 
   -> request_raw #3, retry until success          
```

An architecture which would avoid this issue:

```
[List of URLs] -> [Queue]
   -> request_raw #1, make a single request, if retry is needed, add a retry request to Queue
   -> request_raw #2, make a single request, if retry is needed, add a retry request to Queue
   -> request_raw #3, make a single request, if retry is needed, add a retry request to Queue
```

I think changing it is out of scope here, unless @victor-torres you see an easy way to do it.

victor-torres replied:
No, I think we can leave this change to a queue/scheduler for another opportunity.

ivanprado commented on autoextract/__main__.py:
```diff
@@ -116,6 +118,12 @@ def read_input(input_fp, intype, page_type):
+    p.add_argument("--disable-retries", action="store_true",
+                   help="Disable network, Request-level, and Query-level "
+                        "errors retrying.")
+    p.add_argument("--skip-query-errors", action="store_true",
```
`skip-query-errors` doesn't reflect well what the flag is. Some ideas: `--dont-retry-on-query-errors` , `--never-retry-on-query-errors`, `--no-retry-on-query-errors` `--avoid-retrying-query-errors`, `--disable-query-errors-retrials`. 

Also probably the user of the library won't have any idea what "Query-level error" means at all. Probably we should clarify it somewhere. Maybe here is a good place.

kmike commented on autoextract/aio/client.py:
```diff
@@ -77,19 +172,31 @@ async def request_raw(query: Query,
+
+        post_kwargs = dict(
+            url=endpoint,
+            json=request_processor.pending_queries,
```
My understanding of the batch handling logic:

* Make a batch request
* If there are query-level errors, retry them all in a single batch request. If there is a retry time specified for some of the errors, use max.
* Repeat until there are no errors or retry stop condition is met.
* Return a list with all results, including successes and last received errors (if they're not resolved by retries).

Is it correct? If so, what do you think about adding a comment somewhere, explaining this logic? It is spread over retry policy, RequestProcessor and request_raw function, so it is not straight forward to understand the behavior just from reading code.

victor-torres replied:
I've tried to better explain the retry logic within this commit https://github.com/scrapinghub/scrapinghub-autoextract/pull/7/commits/bc6e032b412bef7d79e04e14e0c5e7f4a567b0b6

victor-torres commented:
@ivanprado @kmike @ejulio 

There you are, guys. I've just refactored the source code to use a max retries integer argument for Query-level errors.
Let me know if it looks good to you and I could proceed to documentation.

ivanprado commented on autoextract/aio/client.py:
```diff
@@ -173,10 +177,17 @@ async def request_raw(query: Query,
         agg_stats = AggStats()  # dummy stats, to simplify code
 
+    if max_query_error_retries and not handle_retries:
+        logger.warning(
```
I saw this message in another part of the code. It will be shown twice when invoking the cmd. Consider leaving only one, probably to show it only in the cmd?

victor-torres replied:
Yes, but don't you want me to leave it in the most inner part of the code so that it's logged whether user is consuming our library API or using our command-line tool?

victor-torres replied:
The `--disable-retries` command-line argument was removed after a comment from @kmike. This log was also removed in the same occasion.

ivanprado replied:
IMO is better not show the warning at all. For a developer who is using the api, it can be handy to have a default `max_query_error_retries` and then use the flag `handle_retries` to enable/disable the whole retries at all. I think is enough to have it documented in the function docstring

kmike replied:
Why is logger.warning used here, not warnings.warn? I.e. why is it better to show this warning every time the function is called, not just once?

victor-torres replied:
No particular reason. I think `warnings.warn` is a better option here. Refactored.

ivanprado commented on autoextract/aio/retry.py:
```diff
@@ -104,6 +117,7 @@ def __init__(self):
         self.stop_on_throttling_error = stop_never
         self.stop_on_network_error = stop_after_delay(15 * 60)
```
I think we should just keep a single stop for these three `network_error`, `server_error` and `query_error`. 

Something like `self.stop_after_some_time =  stop_after_delay(15 * 60)`

As it is now is confusing

victor-torres replied:
Sure, this is a good time to refactor it!

victor-torres replied:
Done!

ivanprado commented on autoextract/aio/retry.py:
```diff
@@ -113,14 +127,28 @@ def __call__(self, retry_state: RetryCallState) -> bool:
             raise RuntimeError("Invalid retry state exception: %s" % exc)
 
 
+def _exception_factory(fut):
```
Not sure about the `QueryRetryError` and this method. It confused me more than help me when reading the code. Another exception to understand `QueryRetryError` which is different than `QueryError`.

victor-torres replied:
This is supposed to differentiate between RetryErrors that happen because of Query-level error versus the ones that happen because of Request-level and Network errors. This was a suggestion from @ejulio to simplify the logic of the request method.

victor-torres replied:
Documenting: we've discussed this comment on another channel and you've agreed that this is a minor thing that could be archived for now. 🤝

kmike commented on autoextract/__main__.py:
```diff
@@ -116,6 +118,12 @@ def read_input(input_fp, intype, page_type):
     p.add_argument("--shuffle", help="Shuffle input URLs", action="store_true")
+    p.add_argument("--disable-retries", action="store_true",
+                   help="Disable network, Request-level, and Query-level "
+                        "errors retrying.")
```
I'm not sure disabling of network retries makes sense. For example, 429 errors are a part of protocol, and should be handled. What do you think about removing this argument? It still should be possible to disable query-level retries by setting `--max-query-error-retries` to 0.

kmike replied:
The intention of handle_retries flag in request_raw was mostly to allow implementing retries externally (e.g. with a queue).

victor-torres replied:
I was just trying to expose arguments supported by the async client API. It could be useful when testing the percentage of those protocol errors, doesn't it?

victor-torres replied:
Makes sense. Removed.

kmike commented on autoextract/aio/client.py:
```diff
@@ -147,6 +270,8 @@ def request_parallel_as_completed(query: Query,
                                   n_conn=1,
                                   agg_stats: AggStats = None,
+                                  handle_retries=True,
+                                  max_query_error_retries=False,
```
shouldn't it be 0, not False?

victor-torres replied:
Sure, thanks for spotting.

victor-torres replied:
Fixed.

victor-torres commented:
@ejulio @ivanprado @kmike 
Review season is open again guys! 
I've added some comments and fixed lastest feedback. 
Let me know if there's something pending.
I believe the PR is ready to be merged.

ivanprado commented:
@victor-torres I'll give another review now, but I think before merging it would be a good idea to do a run as the litmus test. I'll suggest you a URLs dataset.

ivanprado commented on autoextract/aio/client.py:
```diff
@@ -34,11 +37,102 @@ class Result(List[Dict]):
+                    self._enqueue_error(query_result, query_exception)
+                    continue
+
+            self._query_successes.append(query_result)
```
Just to be sure I get it, in `_query_successes` will be also results that generated a `QueryError` when `from_query_result` was invoked, but that are not retriable. I think is right. 

And also never a QueryError exception is raised from `request_raw`

victor-torres replied:
> Just to be sure I get it, in _query_successes will be also results that generated a QueryError when from_query_result was invoked, but that are not retriable. I think is right.

Yes, that's right! I'm renaming those variables to something more meaningful and less confusing: `_complete_queries` and `_retriable_queries`. I've also slightly updated docstrings to reflect this change.

> And also never a QueryError exception is raised from request_raw

Correct. Whenever `request()` raises a Query-level error, it should be captured by Tenacity rules and retried. They're only raised if retries are enabled so...

ivanprado replied:
Nice.

ivanprado commented on autoextract/aio/client.py:
```diff
@@ -61,14 +155,19 @@ async def request_raw(query: Query,
+       if there is a Request-level error returned by the API
+       which is not a throttling response
+       (e.g. it can be raised for incorrect request).
+    2) tenacity.RetryError,
```
Is that right? tenacity.RetryErrors are raised by `request_raw`? I think is not the case, but not sure. I think they shouldn't. Also, I think `RequestError` should also be raised in the case that `handle_retries` is False. Not sure if this the case...

victor-torres replied:
> Is that right? tenacity.RetryErrors are raised by request_raw? I think is not the case, but not sure.

Yes, that's true. See here:

```python
    try:
        # Try to make a batch request
        result = await request()
    except QueryRetryError:
        # If Tenacity fails to retry a QueryError because the max number of
        # retries or a timeout was reached, get latest results combining
        # error and successes and consider it as the final result.
        result = request_processor.get_latest_results()
    except Exception:
        agg_stats.n_fatal_errors += 1
        raise
```

The `tenacity.RetryError` exceptions are captured by the `except Exception` just to increment fatal error stats and then they are raised again. This was also the behavior before my pull request.

> I think they shouldn't. 

This is something we need to discuss, but it's probably out of scope. 
Any thoughts, @kmike and @ejulio?

> Also, I think RequestError should also be raised in the case that handle_retries is False. Not sure if this the case...

It is raised. What will happen is that since retries are disabled, Tenacity won't capture it and it will be re-raised per the `except Exception` block as described above.

```python
                if resp.status >= 400:
                    # ...
                    raise RequestError(
                        request_info=resp.request_info,
                        history=resp.history,
                        status=resp.status,
                        message=resp.reason,
                        headers=resp.headers,
                        response_content=content
                    )
```

ivanprado replied:
> The tenacity.RetryError exceptions are captured by the except Exception just to increment fatal error stats and then they are raised again. This was also the behavior before my pull request.

Ok, I think we have to discuss that, and I think it should be in the scope of this PR. My point is that we should never raise `tenacity.RetryError`. Instead, we should raise the underlying `RequestError` for request errors, or ` request_processor.get_latest_results()` otherwise. We could include the number of retrials in `RequestError` in case we want to make this information accessible by the client.

ivanprado replied:
To be done separately. Created this issue: https://github.com/scrapinghub/scrapinghub-autoextract/issues/11

ivanprado commented on tests/test_aio/test_request_processor.py:
```diff
@@ -0,0 +1,333 @@
+    # Initialize our request processor with this query
+    request_processor = RequestProcessor(query=initial_query)
+
+    # Since we haven't processed any response yet,
```
I like all these comments. Much easy to read the test :-)

victor-torres replied:
Thank you!

victor-torres commented on README.rst:
```diff
@@ -133,6 +133,22 @@ input queries before sending them to the API:
+Retries
+~~~~~~~
+
+By default, the command-line interface will automatically retry Network and
```
Hey, @ivanprado. I've included a section to briefly talk about errors (network, request-level, and query-level) and try to explain that some errors could be retried while others don't. 

I have also included a modified version of your suggested snippet.

victor-torres commented on autoextract/__main__.py:
```diff
@@ -116,6 +117,8 @@ def read_input(input_fp, intype, page_type):
                    help="log level")
     p.add_argument("--shuffle", help="Shuffle input URLs", action="store_true")
+    p.add_argument("--max-query-error-retries", type=int, default=0,
+                   help="Max number of Query-level error retries.")
```
Done.

ivanprado commented on README.rst:
```diff
@@ -133,6 +133,29 @@ input queries before sending them to the API:
+    - Proxy errors
+    - ...
+
+Some errors could be retried while others don't. For example, you could retry
```
Alternative:
`
Some errors could be retried because there are chances the response will change in further requests, but for others, it makes no sense. 
`

victor-torres replied:
I've updated this part os the readme file.

ivanprado commented on autoextract/aio/client.py:
```diff
@@ -52,12 +52,12 @@ def __init__(self, query: Query, max_retries: int = 0):
         self.pending_queries = query_as_dict_list(query)
         self._max_retries = max_retries
-        self._query_successes: List[Dict] = list()
+        self._complete_queries: List[Dict] = list()
```
Good idea the renaming :+1:

victor-torres commented:
Since we've moved the concern with `tenacity.RetryError` to another issue, are you ready to approve this PR, @ivanprado?

ivanprado commented:
>  are you ready to approve this PR, @ivanprado?

From the code perspective, yes, the PR is ready. Great work! :+1: Only this comment from my side remains to be addressed: https://github.com/scrapinghub/scrapinghub-autoextract/pull/7#discussion_r455707736

I'll wait until seeing the results for tests you will run for the final green light to the PR.

kmike commented on README.rst:
```diff
@@ -133,6 +133,49 @@ input queries before sending them to the API:
+
+    python -m autoextract urls.txt --page-type articles --max-query-error-retries 3 --output res.jl
+
+Queries with success will be buffered while failing queries will be retried
+until the max number of retries or a timeout is reached. If it's not possible
```
This explanation sounds a bit confusing in the context of the example above. Successful responses are written to a file, not buffered. Technically what you've written is correct though, as CLI uses batch-size 2 by default - so successful responses can be buffered for some time. But how it actually works is a bit confusing without mentioning batches.

What do you think about wording it in a slightly different way? Something along the lines "Failing queries are retried until ...; if not possible, the last available result is written to the output. When batches are used, successful responses may be buffered in memory until all queries in a batch are resolved, because of the limitations of the current implementation."

victor-torres replied:
I've updated this part os the readme file.

kmike commented on autoextract/aio/client.py:
```diff
@@ -14,8 +15,10 @@
+from .retry import autoextract_retry, QueryRetryError
+from .errors import RequestError, QueryError
+
+logger = logging.getLogger('autoextract')
```
Why do you prefer explicit logger name instead of `logger = logging.getLogger(__name__)` here?

victor-torres replied:
Not a preference. I've probably copied this line from `__main__.py`. 
Changing to `__name__`.

ivanprado commented on README.rst:
```diff
@@ -171,10 +179,15 @@ if you are interested in a higher success rate.
+the last available result is written to the output
+including both queries with success and the ones with errors.
+
+When batches are used,
```
IMO this is an implementation detail. It is a comment I would include the in docstring of the methods, but not in the README. As is, it would be very confusing for the reader. A CMD user wouldn't understand it, and also It wouldn't extract any practical conclusion about how this is affecting him. I propose just to remove the paragraph.

victor-torres replied:
What do you think, @kmike?

kmike replied:
I'm fine with @ivanprado's suggestion. I was on fence between suggesting to remove this section and suggesting to put more effort to make it more precise. Both sound good to me.

victor-torres replied:
Removing then 🚀

ejulio commented on README.rst:
```diff
@@ -133,6 +133,62 @@ input queries before sending them to the API:
+Errors
+~~~~~~
+
+The following errors could happen while making requests:
```
I prefer _fetching_ or _downloading_ instead of _making_ here.
But it's personal

victor-torres replied:
IMO

- you fetch responses, not requests
- you download responses, not requests
- you make/send a request and server answers/sends a response

ejulio commented on README.rst:
```diff
@@ -133,6 +133,62 @@ input queries before sending them to the API:
+    - Proxy errors
+    - ...
+
+Some errors could be retried while others don't.
```
Some errors _can_ be..

victor-torres replied:
Both are modal verbs and I think they have very similar meaning despite slightly different intonation in this case. I'm replacing it anyway because I think it makes the sentence sound more clear and certain.

ejulio commented on README.rst:
```diff
@@ -133,6 +133,62 @@ input queries before sending them to the API:
+Some errors could be retried while others don't.
+
+For example,
+you could retry a query with a Proxy Timeout error
```
If changing the above, then this one could be changed as well, I think..

victor-torres replied:
Done.

ejulio commented on README.rst:
```diff
@@ -133,6 +133,62 @@ input queries before sending them to the API:
+
+By default, we will automatically retry Network and Request-level errors.
+You could also enable Query-level errors retries
+by specifying the ``--max-query-error-retries`` argument.
```
Maybe ``--max-query-error-retries=<N RETRIES>``

victor-torres replied:
This is how arguments are referenced by other parts of the readme file. 
I also think that these arguments are well covered by the command-line utility's help texts.

ejulio commented on autoextract/__main__.py:
```diff
@@ -24,7 +24,7 @@
 
 async def run(query: Query, out, n_conn, batch_size, stop_on_errors=False,
-              api_key=None):
+              api_key=None, max_query_error_retries=0):
```
Since it's already a default value in the CLI, it could be mandatory here.
Unless, the `run` interface needs to have it as optional

victor-torres replied:
I guess it's good for communicating that this is an optional parameter.

kmike commented on autoextract/aio/client.py:
```diff
@@ -114,19 +233,33 @@ async def request():
+        # successfully retry failing requests.
+        #
+        # In addition to handle_retries=True, QueryRetryError also depends on
+        # max_query_error_retries being a non-nullable positive integer.
```
```suggestion
        # max_query_error_retries being greater than 0.
```

victor-torres replied:
That's true 😂

kmike commented on autoextract/aio/client.py:
```diff
@@ -34,11 +37,103 @@ class Result(List[Dict]):
+
+        Return successful queries and also failed ones.
+
+        If `self._handle_exceptions` is True,
```
Is this comment up to date?

victor-torres replied:
It's supposed to be `_max_retries`. Updated.

kmike commented on autoextract/aio/client.py:
```diff
@@ -114,19 +233,33 @@ async def request():
                 # good response
```
it is not necessarily good, as `request_processor.process_results` may also raise an exception

victor-torres replied:
That's true. I'm just removing this comment for now.

kmike commented on autoextract/aio/client.py:
```diff
@@ -179,7 +177,7 @@ async def request_raw(query: Query,
 
     if max_query_error_retries and not handle_retries:
-        logger.warning(
+        warnings.warn(
```
stacklevel=2 could make this warning more useful, as it'd be pointing to the caller code, not inside the function

kmike commented:
Thanks @victor-torres for the implementation, and @ivanprado + @ejulio for the reviews and suggestions!

victor-torres commented:
Thank you for all the time invested in reviewing this code, @kmike, @ivanprado, and @ejulio! 🚀

kmike commented on autoextract/aio/client.py:
```diff
@@ -147,6 +280,8 @@ def request_parallel_as_completed(query: Query,
                                   batch_size=1,
                                   n_conn=1,
                                   agg_stats: AggStats = None,
+                                  handle_retries=True,
```
@victor-torres sorry, I missed this - what's the use case for adding this argument to request_parallel_as_completed? When False is passed, won't a single throttling error cause an exception, and stop the processing?

victor-torres replied:
Just keeping interfaces consistent.

victor-torres replied:
Should I keep it? Let me know and I can remove it in another PR.

kmike replied:
I think it makes sense to remove it - it'd be a new feature, with unclear use case, which can be actually hamful for users (they can turn set it to False without understanding the consequences, and then their crawl stops in the middle, after half an hour).

kmike commented on autoextract/aio/client.py:
```diff
@@ -147,6 +280,8 @@ def request_parallel_as_completed(query: Query,
                                   n_conn=1,
                                   agg_stats: AggStats = None,
+                                  handle_retries=True,
+                                  max_query_error_retries=0,
```
We should document this argument.

victor-torres replied:
I could open a new PR adding docs.
